src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                QueryExpressionProcessingInfo queryExpressionProcessingInfo,
+                out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // The conversion requires yield return which cannot be added to lambdas and anonymous method declarations.
+                if (IsWithinImmediateLambdaOrAnonymousMethod(returnStatement))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                var memberDeclarationNode = FindParentMemberDeclarationNode(returnStatement);
+                if (memberDeclarationNode == null)
+                {
+                    throw new ArgumentException("Must bail out on previous steps");
+                }
+
+                var declaredSymbol = _semanticModel.GetDeclaredSymbol(memberDeclarationNode);
+                ITypeSymbol typeSymbol;
+                switch (declaredSymbol.Kind)
+                {
+                    case SymbolKind.Local:
+                        typeSymbol = ((ILocalSymbol)declaredSymbol).Type;
+                        break;
+                    case SymbolKind.Method:
+                        typeSymbol = ((IMethodSymbol)declaredSymbol).ReturnType;
+                        break;
+                    case SymbolKind.Property:
+                        typeSymbol = ((IPropertySymbol)declaredSymbol).Type;
+                        break;
+                    default:
+                        throw new ArgumentException(declaredSymbol.Kind.ToString());
+                }
+
+                if (typeSymbol.OriginalDefinition?.SpecialType != SpecialType.System_Collections_Generic_IEnumerable_T)
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // before: return from a in b select a;
+                // after: 
+                // foreach(var a in b)
+                // {
+                //      yield return a;
+                // }
+                var statements = GenerateStatements((ExpressionSyntax expression)
+                    => SyntaxFactory.YieldStatement(SyntaxKind.YieldReturnStatement, expression), queryExpressionProcessingInfo);
+
+
+                // if there are more than one return in the method, add yield break;
+                if (memberDeclarationNode.DescendantNodes().OfType<ReturnStatementSyntax>().Count() != 1)
+                {
+                    var yieldBreakStatement = SyntaxFactory.YieldStatement(SyntaxKind.YieldBreakStatement);
+                    documentUpdateInfo = new DocumentUpdateInfo(returnStatement, statements.Concat(new[] { yieldBreakStatement }));
+                }
+                else
+                {
+                    documentUpdateInfo = new DocumentUpdateInfo(returnStatement, statements);
+                }
+
+                return true;
+            }
+
+            private static SyntaxNode FindParentMemberDeclarationNode(SyntaxNode node)
```

Comment:
>FindParentMemberDeclarationNode [](start = 38, length = 31)

It doesn't look like this function is going to properly handle property and event accessors, operators, etc. We probably could rely on SemanticModel.GetEnclosingSymbol API to get to the symbol directly, but then we would need to find a reliable way to get the declaration syntax for that, perhaps DeclaringSyntaxReferences can help.